build-system/tasks/generate-runner.js

```diff
@@ -0,0 +1,129 @@
+const {isTravisBuild} = require('../travis');
+
+const antExecutable = 'third_party/ant/bin/ant';
+const runnerDir = 'build-system/runner';
+const buildFile = `${runnerDir}/build.xml`;
+const runnerDistDir = `${runnerDir}/dist`;
+const generatedAtCommitFile = 'GENERATED_AT_COMMIT';
+
+/**
+ * Determines if runner.jar needs to be regenerated
+ * @param {string} subDir
+ * @return {boolean}
+ */
+function shouldGenerateRunner(subDir) {
+  const runnerJarDir = `${runnerDistDir}/${subDir}`;
+  const runnerJar = `${runnerJarDir}/runner.jar`;
+
+  // Always generate on Travis
+  if (isTravisBuild()) {
+    return true;
+  }
+
+  // The binary hasn't been generated
+  if (!fs.existsSync(runnerJar)) {
+    return true;
+  }
+
+  // The binary was generated at a different commit
+  const currentCommit = gitCommitHash();
+  const generatedAtCommit = fs
+    .readFileSync(`${runnerJarDir}/${generatedAtCommitFile}`, 'utf8')
+    .toString();
+  if (currentCommit != generatedAtCommit) {
+    return true;
+  }
+
+  // There are local changes in the build-system/runner directory
+  const localRunnerChanges = gitDiffPath(runnerDir, gitCommitHash());
+  if (localRunnerChanges) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Generates runner.jar if required
+ * @param {string} subDir
+ */
+async function maybeGenerateRunner(subDir) {
+  if (shouldGenerateRunner(subDir)) {
+    await generateRunner(subDir);
+  }
+}
+
+/**
+ * Writes a file to the runner dir to indicate when it was last generated
+ * @param {string} runnerJarDir
+ */
+function writeGeneratedAtCommitFile(runnerJarDir) {
+  let generatedAtCommit = gitCommitHash();
+  const localRunnerChanges = gitDiffPath(runnerDir, gitCommitHash());
+  if (localRunnerChanges) {
+    generatedAtCommit += '+local';
+  }
+  fs.ensureDirSync(runnerJarDir);
+  fs.writeFileSync(
+    `${runnerJarDir}/${generatedAtCommitFile}`,
+    generatedAtCommit
+  );
+}
+
+/**
+ * Generates the custom closure compiler binary (runner.jar) and drops it in the
+ * given subdirectory of build-system/runner/dist/ (to enable concurrent usage)
+ * @param {string} subDir
+ *
+ */
+async function generateRunner(subDir) {
+  const generateCmd = `${antExecutable} -buildfile ${buildFile} -Ddist.dir dist/${subDir} jar`;
+  const runnerJarDir = `${runnerDistDir}/${subDir}`;
+  writeGeneratedAtCommitFile(runnerJarDir);
+  const result = getOutput(generateCmd);
+  if (result.stderr) {
+    log(
+      red('ERROR:'),
+      'Could not generate custom closure compiler',
+      cyan(`${runnerJarDir}/runner.jar`)
+    );
+    console.error(red(result.stdout), red(result.stderr));
+    const reason = new Error('Compiler generation failed');
+    reason.showStack = false;
+    return Promise.reject(reason);
+  } else {
+    if (!isTravisBuild()) {
+      log(
+        'Generated custom closure compiler',
+        cyan(`${runnerJarDir}/runner.jar`)
+      );
+    }
+  }
+  return Promise.resolve();
```

Comment:
don't believe you need this since this is an async function

Reply:
This is a companion to line 115:

https://github.com/ampproject/amphtml/blob/84142ec79c283792ad762af250acd390accf5d97/build-system/tasks/generate-runner.js#L115

I prefer to be explicit and return a Promise in both the success and failure cases.

Reply:
Async implies a promise, though. This creates a extra object, and forces an extra 3 queue ticks to move to the next step.

Reply:
Removed in https://github.com/ampproject/amphtml/pull/23434/commits/bf08e7b6eefe82a7f63e164ef9076ad47487ab13.